### PR TITLE
Adapting hlo_legalize_to_lhlo for using buffer assignment

### PR DESCRIPTION
**Pull request by ehsan-nt**
In this PR, we implemented our proposed [BufferAssignment](https://github.com/tensorflow/tensorflow/pull/37212) in HLO-to-LHLO legalization pipeline. Please note that this PR must be merged after BufferAssignment PR since it cannot be built.

### CODE REVIEW

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 1]
```diff
@@ -105,43 +106,41 @@ Value InsertAllocAndDealloc(Location loc, Value result,
...
 
 template <typename HloOpTy, typename LhloOpTy>
-class HloToLhloOpConverter : public ConversionPattern {
+class HloToLhloOpConverter
+    : public xla::BufferAssignmentOpConversionPattern<HloOpTy> {
  public:
-  explicit HloToLhloOpConverter(MLIRContext* context)
-      : ConversionPattern(HloOpTy::getOperationName(), 1, context) {}
-
+  using xla::BufferAssignmentOpConversionPattern<
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: This seems oddly formatted.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 2]
```diff
@@ -409,37 +381,31 @@ class StdToLhloReturnOpConverter : public OpConversionPattern<mlir::ReturnOp> {
...
-      auto dealloc = FindInsertionPointForCopy(operand.value());
-
-      if (dealloc == nullptr) {
-        returnOp.emitOpError()
-            << "Missing dealloc for operand " << operand.index();
-        return matchFailure();
-      }
-      OpBuilder::InsertionGuard guard(rewriter);
-      rewriter.setInsertionPoint(dealloc);
+      rewriter.setInsertionPointToEnd(block);
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Why is this needed?

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 3]
```diff
@@ -409,37 +381,31 @@ class StdToLhloReturnOpConverter : public OpConversionPattern<mlir::ReturnOp> {
...
       rewriter.create<xla_lhlo::CopyOp>(loc, llvm::None, operand.value(),
                                         funcOp.getArgument(returnArgNumber));
     }
-    rewriter.replaceOpWithNewOp<xla_lhlo::TerminatorOp>(returnOp);
+    rewriter.setInsertionPointToEnd(block);
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Why can this no longer use `replaceOpWithNewOp`?

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 4]
```diff
@@ -93,6 +93,7 @@ Value InsertDynamicAllocAndDealloc(Location loc, Value result,
 }
 
 Value InsertAllocAndDealloc(Location loc, Value result,
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Please rename this function, as it no longer inserts dealloc operations.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 5]
```diff
@@ -105,43 +106,41 @@ Value InsertAllocAndDealloc(Location loc, Value result,
...
       ConversionPatternRewriter& rewriter) const final {
+    Operation* op = hloOp.getOperation();
     const auto& original_results = op->getResults();
     SmallVector<Value, 4> buffer_args(operands.begin(), operands.end());
     for (auto result : llvm::enumerate(original_results)) {
       RankedTensorType resultType =
           result.value().getType().dyn_cast<RankedTensorType>();
       if (!resultType) {
-        return matchFailure();
+        return super::matchFailure();
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: As this is static, `ConversionPattern::` would do, too.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc (hunk 4 above)
Comment (sherhut) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Also, you need to fix the `Dynamic...` version of this function, too.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 6]
```diff
@@ -324,8 +329,8 @@ class HloToLhloTensorStoreOpConverter : public ConversionPattern {
 //   "xla_lhlo.terminator"() : () -> ()
 // }
 
-struct HloLegalizeToLhlo : public ModulePass<HloLegalizeToLhlo> {
-  void runOnModule() override {
+struct HloLegalizeToLhlo : public FunctionPass<HloLegalizeToLhlo> {
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Does this need to be a `FunctionPass` or could this be a generic OperationPass?

**Review by sherhut — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
@dfki-ehna Can you please resolve conflicts? Thanks!

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc (hunk 2 above)
Comment (dfki-ehna) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Resolved. This conversion pattern is going to be moved to [Buffer Assignment ](https://github.com/tensorflow/tensorflow/pull/37212) and changed to a template pattern converter.

```c++
template <typename ReturnOpSourceTy, typename ReturnOpTargetTy,
          typename CopyOpTy>
class NonVoidToVoidReturnOpConversion
    : public OpConversionPattern<ReturnOpSourceTy> {
```

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc (hunk 4 above)
Comment (dfki-ehna) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Renaming is done. The current [BufferAssignment](https://github.com/tensorflow/tensorflow/pull/37212) doesn't support dynamically shaped types but in the near future, it will support it. If we use it now, the AllocOp positions would be wrong. Should we still update this function?

Comment (sherhut) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: No, we rely on this working. But supporting dynamic shapes should be high on the feature list.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 7]
```diff
@@ -338,69 +334,28 @@ struct HloLegalizeToLhlo : public ModulePass<HloLegalizeToLhlo> {
...
-      rewriter.applySignatureConversion(&funcOp.getBody(), conversion);
-    });
-    return success();
-  }
-};
-
 /// Transforms ReturnOp to LhloTerminator. CopyOp is inserted to copy each
 /// result to the corresponding buffer argument.
-class StdToLhloReturnOpConverter : public OpConversionPattern<mlir::ReturnOp> {
+class StdToLhloReturnOpConverter
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Should this make use of the pattern provided by the buffer allocation header? Also fine in a follow up CL.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 8]
```diff
@@ -465,15 +413,15 @@ void populateHLOToLHLOConversionPattern(MLIRContext* context,
       HloToLhloOpConverter<xla_hlo::SqrtOp>,
       HloToLhloOpConverter<xla_hlo::SubOp>,
       HloToLhloOpConverter<xla_hlo::TanhOp>,
-      HloToLhloReduceOpConverter,
+      HloToLhloReduceOpConverter,
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Please format.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 9]
```diff
@@ -230,26 +224,28 @@ struct HloToLhloReduceOpConverter
...
 };
 
 // TODO(b/137624192): Rewrite into a copy and elide copy if possible.
-class HloToLhloTensorStoreOpConverter : public ConversionPattern {
+class HloToLhloTensorStoreOpConverter
+    : public xla::BufferAssignmentOpConversionPattern<mlir::TensorStoreOp> {
  public:
-  explicit HloToLhloTensorStoreOpConverter(MLIRContext* context)
-      : ConversionPattern(TensorStoreOp::getOperationName(), 1, context) {}
+  using xla::BufferAssignmentOpConversionPattern<
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Please format. We typically use `clang-format` and git has a `git clang-format` helper that automates this.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 10]
```diff
@@ -105,38 +97,34 @@ Value InsertAllocAndDealloc(Location loc, Value result,
...
       if (!resultType) {
-        return failure();
+        return ConversionPattern::failure();
       }
       if (resultType.hasStaticShape()) {
-        buffer_args.push_back(
-            InsertAllocAndDealloc(op->getLoc(), result.value(), &rewriter));
+        buffer_args.push_back(InsertAlloc(
+            op->getLoc(), result.value(),
+            xla::BufferAssignmentOpConversionPattern<HloOpTy>::bufferAssignment,
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Can this be `this->bufferAssignment`?

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc (hunk 9 above)
Comment (sherhut) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Just realized this might be formatted correctly and just look weird. In that case please ignore my comment.

**Review by sherhut — CHANGES_REQUESTED** [review]
Please address formatting nits and then we can land this.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc (hunk 3 above)
Comment (dfki-ehna) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: The generic version of this conversion pattern moved to [BufferAssignment](https://github.com/tensorflow/tensorflow/pull/37212). Your suggestion has also been addressed.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc (hunk 7 above)
Comment (dfki-ehna) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Yes, and it's done.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc (hunk 9 above)
Comment (dfki-ehna) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Yes, the clang-format formatted the code like that.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc (hunk 10 above)
Comment (dfki-ehna) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Done.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc [hunk 11]
```diff
@@ -92,40 +83,33 @@ Value InsertDynamicAllocAndDealloc(Location loc, Value result,
...
   }
   auto memref_type =
       MemRefType::get(result_type.getShape(), result_type.getElementType());
-
-  Operation* op = result.getDefiningOp();
-  auto block = op->getBlock();
-
-  OpBuilder allocBuilder(op);
-  allocBuilder.setInsertionPointToStart(block);  // Inserting at the beginning
+  auto allocPosition = bufferAssignment->computeAllocPosition(result);
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc:96]: Looking at this I realized that this is used by rewrite patterns, so it should not use a builder of its own. Instead, use the `rewriter`. It implements the same interface, so you can also call `restoreInsertionPoint` on it.

Comment (sherhut) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc:96]: You should also use an `InsertionGuard` to make sure that your changing of the position does not leak into the calling context. Something like 'OpBuilder::InsertionGuard(rewriter)` at the top should do.

**Review by sherhut — DISMISSED** [review]
Please do the fix in a follow up.

File: tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc (hunk 11 above)
Comment (dfki-ehna) [reply tensorflow/compiler/mlir/xla/transforms/hlo_legalize_to_lhlo.cc]: Done.

**Review by sherhut — DISMISSED** [review]

**Comment by gbaned** [comment]
@dfki-ehna Can you please check build failures. Thanks!

File: tensorflow/compiler/mlir/xla/BUILD [hunk 12]
```diff
@@ -317,6 +317,7 @@ cc_library(
     srcs = ["transforms/hlo_legalize_to_lhlo.cc"],
     deps = [
         ":hlo",
+        ":buffer_assignment",
```
Comment (sherhut) [inline tensorflow/compiler/mlir/xla/BUILD:320]: This is not sorted.

**Review by sherhut — DISMISSED** [review]

**Review by pifon2a — APPROVED** [review]

File: tensorflow/compiler/mlir/xla/transforms/rewriters.h [hunk 13]
```diff
@@ -38,8 +41,9 @@ void PopulateXlaToStdPatterns(OwningRewritePatternList *patterns,
                               MLIRContext *ctx);
 
 // Collection of rewrite patterns for lowering of HLO to LHLO dialect.
-void populateHLOToLHLOConversionPattern(MLIRContext *context,
-                                        OwningRewritePatternList *patterns);
+void populateHLOToLHLOConversionPattern(
```
Comment (rthadur) [inline tensorflow/compiler/mlir/xla/transforms/rewriters.h:44]: `third_party/tf_runtime/tools/tf_kernel_gen/kernel_lowering.cc:98:72: error: too few arguments to function call, expected 3, have 2
    ::mlir::xla_hlo::populateHLOToLHLOConversionPattern(&ctx, &patterns);
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                ^
./third_party/tensorflow/compiler/mlir/xla/transforms/rewriters.h:44:6: note: 'populateHLOToLHLOConversionPattern' declared here
void populateHLOToLHLOConversionPattern(`

Comment (rthadur) [reply tensorflow/compiler/mlir/xla/transforms/rewriters.h:44]: @dfki-ehna can you please check above error ?

**Review by rthadur — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/mlir/xla/transforms/rewriters.h (hunk 13 above)
Comment (dfki-ehna) [reply tensorflow/compiler/mlir/xla/transforms/rewriters.h:44]: @rthadur we have fixed this error locally. However, [Buffer Assignment](https://github.com/tensorflow/tensorflow/pull/37212) has been reverted. We are opening a new PR for Buffer Assignment and adding new features and tests. We should wait for getting it landed again.

Comment (rthadur) [reply tensorflow/compiler/mlir/xla/transforms/rewriters.h:44]: thank you, do you have a PR number ?

Comment (sherhut) [reply tensorflow/compiler/mlir/xla/transforms/rewriters.h:44]: I don't think it was reverted.

**Comment by gbaned** [comment]
@dfki-ehna Can you please resolve conflicts? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
